src/eduid_webapp/security/views/webauthn.py

```diff
@@ -0,0 +1,165 @@
+from fido2.ctap2 import AttestationObject
+from fido2 import cbor
+from fido2.utils import websafe_encode
+from fido2.ctap2 import AttestedCredentialData
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from OpenSSL import crypto
+
+from eduid_userdb.credentials import Webauthn
+from eduid_userdb.security import SecurityUser
+from eduid_common.api.decorators import require_user, MarshalWith, UnmarshalWith
+from eduid_common.api.utils import save_and_sync_user
+from eduid_webapp.security.helpers import credentials_to_registered_keys, compile_credential_list
+from eduid_webapp.security.schemas import WebauthnOptionsResponseSchema, WebauthnRegisterRequestSchema
+from eduid_webapp.security.schemas import SecurityResponseSchema, RemoveWebauthnTokenRequestSchema
+from eduid_webapp.security.schemas import VerifyWithWebauthnTokenRequestSchema
+from eduid_webapp.security.schemas import VerifyWithWebauthnTokenResponseSchema
+
+
+WEBAUTHN_SERVER = None
+
+def update_webauthn_server(rp_id, name='eduID security API'):
+    rp = RelyingParty(rp_id, name)
+    server = Fido2Server(rp)
+    global WEBAUTHN_SERVER
+    WEBAUTHN_SERVER = server
+    return server
+
+def get_webauthn_server():
+    if WEBAUTHN_SERVER is not None:
+        return WEBAUTHN_SERVER
+    return update_webauthn_server(current_app.config['FIDO2_RP_ID'])
+
+
+def make_credentials(creds):
+    credentials = []
+    for cred in creds:
+        cred_data = base64.b64decode(cred.credential_data.encode('ascii'))
+        credential_data, rest = AttestedCredentialData.unpack_from(cred_data)
+        if rest:
+            continue
+        credentials.append(credential_data)
+    return credentials
+
+
+webauthn_views = Blueprint('webauthn', __name__, url_prefix='/webauthn', template_folder='templates')
+
+@webauthn_views.route('/register/begin', methods=['GET'])
+@MarshalWith(WebauthnOptionsResponseSchema)
+@require_user
+def registration_begin(user):
+    user_webauthn_tokens = user.credentials.filter(Webauthn)
+    if user_webauthn_tokens.count >= current_app.config['WEBAUTHN_MAX_ALLOWED_TOKENS']:
+        current_app.logger.error('User tried to register more than {} tokens.'.format(
+            current_app.config['WEBAUTHN_MAX_ALLOWED_TOKENS']))
+        resp = {'_status': 'error', 'message': 'security.webauthn.max_allowed_tokens'}
+        cbor_resp = cbor.dumps(resp)
+        return Response(response=cbor_resp, status=200, mimetype='application/cbor')
+    creds = make_credentials(user_webauthn_tokens.to_list())
+    server = get_webauthn_server()
+    if user.surname is None or user.display_name is None:
+        resp = {'_status': 'error', 'message': 'security.webauthn-missing-pdata'}
+        cbor_resp = cbor.dumps(resp)
+        return Response(response=cbor_resp, status=200, mimetype='application/cbor')
+    registration_data, state = server.register_begin({
+        'id': str(user.user_id).encode('ascii'),
```

Comment:
Do we not want to use eppn instead of the mongo _id?

Reply:
yes - changing it